### PR TITLE
[#658] update subjects via CLI + keep trace

### PR DESCRIPTION
**Pull request by fenekku**
- closes https://github.com/galterlibrary/Prism/issues/658

### CODE REVIEW

File: galter_subjects_utils/cli.py [hunk 1]
```diff
@@ -79,3 +81,33 @@ def lcsh(**parameters):
...
+def update_subjects(**parameters):
+    """Update subjects in running instance according to deltas file."""
+    deltas = [d for d in read_csv(parameters["deltas_file"])]
+    logger = SubjectDeltaLogger(filepath=parameters["output_file"])
+    updater = SubjectDeltaUpdater(deltas, logger)
+    keep_trace = KeepTrace(
+        field=parameters.get("keep_trace_field") or None,
+        template=parameters.get("keep_trace_template") or None
+    )
+    updater.update(keep_trace)
```
Comment (fenekku) [inline galter_subjects_utils/cli.py:112]: TBH `keep_trace` could be passed as a constructor argument. For these big one pass algorithms, OO is very weird. Thoughts on this welcome!

File: galter_subjects_utils/updater.py [hunk 2]
```diff
@@ -22,78 +26,114 @@ def filter_ops_by_type(ops_data, _type):
...
+    Return (-1, {}) if not found.
+    """
+    return next(
+        ((i, s) for i, s in enumerate(subjects) if s.get("id") == id_),
+        (-1, {})
+    )
+
 
-    This is actually a rare op, because usually subjects are replaced.
+def op_replace(subjects, op_data):
```
Comment (fenekku) [inline galter_subjects_utils/updater.py:40]: Just moving definition of `op_replace` above `op_remove` to mimic their order in `apply_op_data_change`

File: tests/test_updater.py [hunk 3]
```diff
@@ -153,11 +157,36 @@ def update_result(
 
 
 # Test utilities
-def any_contains(dict_, dicts):
+def any_contains(dicts, dict_):
```
Comment (fenekku) [inline tests/test_updater.py:160]: Simple refactor that better follows reading order.

File: tests/test_updater.py [hunk 4]
```diff
@@ -153,11 +157,36 @@ def update_result(
...
 
+def get_subjects_of_record_from_db(pid):
+    """Return DB subjects of record with `pid`."""
+    records_service = current_service_registry.get("records")
+    record_result = records_service.read(system_identity, pid)
+    record_dict = record_result.to_dict()
+    return record_dict["metadata"]["subjects"]
+
+
+def get_records_from_de():
```
Comment (fenekku) [inline tests/test_updater.py:173]: DE = document engine

File: tests/test_updater.py [hunk 5]
```diff
@@ -283,3 +308,121 @@ def test_update_logging(update_result, delta_logger, records_data_w_subjects):
     assert "" == r1_log_entry["error"]
     r1_delta = "http://example.org/foo/0 -> X"
     assert r1_delta == r1_log_entry["deltas"]
+
+
+def test_update_keep_trace(
+    create_subject_data, minimal_record_input, create_record_data_fn,
+):
+    # Test that keep_trace template is placed in keep_trace field
+    # if subject delta op is marked as keep_trace
```
Comment (fenekku) [inline tests/test_updater.py:317]: In future PR we'll settle what values of keep_trace in the csv will be truthy / what values will be falsy.

**Review by carsonicator — APPROVED** [review]
Thanks @fenekku

**Review by Meowcenary — APPROVED** [review]
This is a lot to look at as a whole so I tried to focus on the smaller parts that made more sense to me. I'm going to approve since i'm at the end of my day and have been reading through this for awhile, but I'll plan on taking another look tomorrow with fresh eyes just to be thorough.

File: galter_subjects_utils/cli.py (hunk 1 above)
Comment (fenekku) [reply galter_subjects_utils/cli.py:112]: In a way I put it there to elicit some response discussion ;). For these kinds of algorithms, I am having a hard time seeing why methods should take an argument when it really operates as loading up the constructor and firing it off with a method call. But I don't know that I have a solid argument against it either... As you say, keeping with the approach so far is a good enough argument.  I will put it back in the constructor.

I was thinking a lot about OO structure here and I think the bottom line for this kind of algorithm is that it's easier to test / there no globally accessible state. Other OO stuff doesn't really matter here.

**Comment by fenekku** [comment]
Just switched the keep_trace back into the constructor. I will keep this around until close to lunch before merging.

File: galter_subjects_utils/updater.py [hunk 6]
```diff
@@ -152,13 +198,58 @@ def at_least_1_subject_targeted(ops_data):
...
+        """Assign expanded template."""
+        final_key = self.field.split(".")[-1]
+        dict_[final_key] = self.template.format(subject=subject)
+
+
 class SubjectDeltaUpdater:
     """Translates delta operations into actual changes."""
 
-    def __init__(self, ops_data, logger):
+    def __init__(self, ops_data, logger, keep_trace):
```
Comment (Meowcenary) [inline galter_subjects_utils/updater.py:248]: I know it's a small change, but seeing the change in full I think this was the right choice

**Review by Meowcenary — APPROVED** [review]
Gave it another look today and I still got nothing to add. Looks good!

**Pull request merged**